std/experimental/color.d

```diff
@@ -0,0 +1,704 @@
+    enum string components = components_;
+    enum ColorSpace colorSpace = colorSpace_;
+
+    // various useful introspection
+    enum bool hasAlpha = components.canFind('a');
+
+    // each color space has a slightly different constructor and some contractual requirements
+    static if(colorSpace == ColorSpace.XYZ)
+    {
+        static assert(components.allIn("xyza"), "Invalid Color component '"d ~ components.notIn("xyza") ~ "'. XYZ colors may only contain components: x, y, z, a"d);
+        static assert(components.canFind('x') && components.canFind('y') && components.canFind('z'), "XYZ colors must have all components: x, y, z"d);
+
+        auto tristimulus() { return tuple(x, y, z); }
+
+        // CIE XYZ constructor
+        this(ComponentType x, ComponentType y, ComponentType z, ComponentType a = 0)
+        {
+            this.x = x;
+            this.y = y;
+            this.z = z;
+            static if(components.canFind('a'))
+                this.a = a;
+        }
+    }
+    static if(isRGB!colorSpace)
+    {
+        // RGB colors may only contain components 'rgb', or 'l' (luminance)
+        // They may also optionally contain an 'a' (alpha) component, and 'x' (unused) components
+
+        // Missing components will be fabricated to produce appropriate values
+
+        static assert(components.allIn("rgblax"), "Invalid Color component '"d ~ components.notIn("rgblax") ~ "'. RGB colors may only contain components: r, g, b, l, a, x"d);
+        static assert(components.anyIn("rgbal"), "RGB colors must contain at least one component of r, g, b, l, a.");
+        static assert(!components.canFind('l') || !components.anyIn("rgb"), "RGB colors may not contain rgb AND luminance components together.");
+
+        // missing components will be available, default to 0
+        static if(!components.canFind('r'))
+        {
+            static if(!components.canFind('l'))
+                enum ComponentType r = 0;
+            else
+                alias r = l;
+        }
+        static if(!components.canFind('g'))
+        {
+            static if(!components.canFind('l'))
+                enum ComponentType g = 0;
+            else
+                alias g = l;
+        }
+        static if(!components.canFind('b'))
+        {
+            static if(!components.canFind('l'))
+                enum ComponentType b = 0;
+            else
+                alias b = l;
+        }
+        static if(!components.canFind('l'))
+            @property ComponentType l() const { return toGrayscale!colorSpace(r, g, b); }
+        static if(!components.canFind('a'))
+            enum ComponentType a = 0;
+
+        static if(components.canFind('l'))
+            auto tristimulus() { auto t = l; return tuple(t, t, t); }
+        else
+            auto tristimulus() { return tuple(r, g, b); }
+
+        // RGB/A constructor
+        this(ComponentType r, ComponentType g, ComponentType b, ComponentType a = 0)
```

Comment:
I think such things should be fail-fast with optional shims that mask away errors.